server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -284,72 +292,58 @@ protected void doXContentBody(XContentBuilder builder, boolean includeDefaults,
-                mapperBuilders.put(builder.name(), builder);
+            public Builder add(FieldMapper.Builder builder) {
+                mapperBuilders.put(builder.name(), builder::build);
                 return this;
             }
 
-            public Builder add(Mapper mapper) {
-                mapperBuilders.put(mapper.simpleName(), new Mapper.Builder(mapper.simpleName()) {
-                    @Override
-                    public Mapper build(BuilderContext context) {
-                        return mapper;
-                    }
-                });
+            public Builder add(FieldMapper mapper) {
+                mapperBuilders.put(mapper.simpleName(), context -> mapper);
                 return this;
             }
 
-            public Builder update(Mapper toMerge, ContentPath contentPath) {
+            public Builder update(FieldMapper toMerge, ContentPath contentPath) {
                 if (mapperBuilders.containsKey(toMerge.simpleName()) == false) {
                     add(toMerge);
                 } else {
-                    Mapper.Builder builder = mapperBuilders.get(toMerge.simpleName());
-                    Mapper existing = builder.build(new BuilderContext(Settings.EMPTY, contentPath));
+                    FieldMapper existing
+                        = mapperBuilders.get(toMerge.simpleName()).apply(new BuilderContext(Settings.EMPTY, contentPath));
                     add(existing.merge(toMerge));
                 }
                 return this;
             }
 
-            @SuppressWarnings("unchecked")
             public MultiFields build(Mapper.Builder mainFieldBuilder, BuilderContext context) {
                 if (mapperBuilders.isEmpty()) {
                     return empty();
                 } else {
+                    Map<String, FieldMapper> mappers = new HashMap<>();
                     context.path().add(mainFieldBuilder.name());
-                    ImmutableOpenMap.Builder mapperBuilders = this.mapperBuilders;
-                    for (ObjectObjectCursor<String, Mapper.Builder> cursor : this.mapperBuilders) {
-                        String key = cursor.key;
-                        Mapper.Builder value = cursor.value;
-                        Mapper mapper = value.build(context);
-                        assert mapper instanceof FieldMapper;
-                        mapperBuilders.put(key, mapper);
+                    for (Map.Entry<String, Function<BuilderContext, FieldMapper>> entry : this.mapperBuilders.entrySet()) {
+                        String key = entry.getKey();
+                        FieldMapper mapper = entry.getValue().apply(context);
+                        mappers.put(key, mapper);
                     }
                     context.path().remove();
-                    ImmutableOpenMap.Builder<String, FieldMapper> mappers = mapperBuilders.cast();
-                    return new MultiFields(mappers.build());
+                    return new MultiFields(mappers);
```

Comment:
Wrap in `unmodifiableMap` or call `Map.of` on it? Just for extra double paranoia that it is immutable.